### PR TITLE
Suggest frequently issued commands for which the user may want to make aliases

### PR DESCRIPTION
**Pull request by 15joeybloom**
I would like some advice on how to write some tests for this code and how to fix the code duplication between SuggestAlias.hs and Recent.hs.

### CODE REVIEW

**Comment by NorfairKing** [comment]
Great work. Some thoughts:

- I think I'd prefer if this would look at all history and not just recent history.
- This looks at the entire commands, and not at word prefixes.
  If we see `stack install X` and `stack install Y`, we may want to alias `stack install`, even if each `stack install Z` doesn't occur a lot.
- I'd like to see a test as well.

Nice-to-haves, but not required or this PR:
- Choose how long of the history to look at.
- It would be nice if this could figure out not to show things that are already aliased.(9/10 of the current items I see are already aliases)

**Comment by 15joeybloom** [comment]
Sounds good! I am currently working on looking at word prefixes of commands and testing. I should have some more work to show soon.

File: hastory-test-utils/test/HastorySpec.hs [hunk 1]
```diff
@@ -6,13 +6,32 @@ module HastorySpec
...
 spec :: Spec
-spec =
+spec = do
     describe "Entry" $ do
         eqSpec @Entry
         jsonSpecOnValid @Entry
         arbitrarySpec @Entry
         genValiditySpec @Entry
+    describe "commandPrefixes" $ do
+        it "Lists only prefixes of the command, ignoring whitespace" $
```
Comment (NorfairKing) [inline hastory-test-utils/test/HastorySpec.hs]: nit: the `it` function is supposed to make a sentence, so you usually don't start the string with a capital.

File: hastory-test-utils/test/HastorySpec.hs [hunk 2]
```diff
@@ -6,13 +6,32 @@ module HastorySpec
...
+            forAll genValid $ \e ->
```
Comment (NorfairKing) [inline hastory-test-utils/test/HastorySpec.hs]: `forAll genValid` should usually be replaced with `forAllValid`, so you get the shrinking as well.

File: hastory/src/Hastory/SuggestAlias.hs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    tups = aggregateSuggestions prefixes
+
+isCDEntry :: Entry -> Bool
+isCDEntry = T.isPrefixOf (T.pack "cd ") . entryText
+
+-- tailSafe drops the empty string
+commandPrefixes :: Entry -> [[Text]]
+commandPrefixes = tailSafe . inits . T.words . entryText
+
+aggregateSuggestions :: [[Text]] -> [([Text], Double)]
```
Comment (NorfairKing) [inline hastory/src/Hastory/SuggestAlias.hs]: Couldn't you replace this `[Text]` by `a`?
If you can, it's probably better to do that.
Parametricity gives you a bunch of safety for free.

File: hastory/src/Hastory/SuggestAlias.hs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+aggregateSuggestions =
+    reverse . sortOn snd . HM.toList . doCountsWith id (const 1.0)
```
Comment (NorfairKing) [inline hastory/src/Hastory/SuggestAlias.hs]: Nit: `reverse` is quite expensive. You could just sort backward with `Down . snd` instead.

**Pull request merged**